sql: unify implicitTxn refs and unify Get interface

couchand opened the pull request:
This PR finishes up the last items that were identified during development of #16746 but left off of that.  The first commit unifies all references to `implicitTxn` to use the flag from `txnState` rather than passing it around separately.  The second commit updates `Get` for session vars to use just the session and not a whole planner.

couchand committed:
sql: pass session not planner in variable Get

This completes the refactoring in #16746, updating Get to take just
the session rather than the planner.

couchand committed:
sql: use implicitTxn only from session.TxnState

This is a preparatory refactor to enable changes to the var interface.
Since #16253, the implicitTxn flag has been maintained on the txnState
struct inside session.  This commit unifies all references to
implicitTxn to use that instead of passing it separately.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16778)
<!-- Reviewable:end -->

andreimatei commented:
LGTM for the 1st commit


nit: the commit ref in the 2nd commit is stale; just say "the previous commit". Even the non-stale ref in the first commit is not particularly useful.

Did anything in particular motivate the 2nd commit? I think it's quite common to pass the `planner` around. For example, it carries the `EvalCtx`. Ideally that'd be the only thing that we pass around, not the `Session`.
For example,  `p.LeaseMgr()` read a lot better than `session.tables.leaseMgr` (the fact that `p.LeaseMgr()` internally uses `p.session.tables.leaseMgr() is an implementation detail and, frankly, horrible).

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/vars.go, line 312 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16778#-KnkztuETIExpIDvV2pe:-KnkztuETIExpIDvV2pf:b-malgu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf8074a0ff54c5e84ca54f6159d5f7d44ff36a04/pkg/sql/vars.go#L312)):*
> ```Go
> 
> 	`transaction isolation level`: {
> 		// TODO(couchand): determine if we need to lock here (is this the session's main goroutine?)
> ```

at least when these functions are called directly by the statement interpreter, I think they'd be indeed be called on the "main goroutine". I don't know if there's more funky calls to these.
Anyway, since nobody cares about their performance, I'd lock the mutex and not worry about it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16778#-:-KnkyLVlTituxffxqKl3:bos4ue1)*
<!-- Sent from Reviewable.io -->

knz commented:
I disagree that the commit ref in the 1st commit is "not useful". We have precedent for that and I appriciate the aid to code navigation.

However indeed the 2nd commit is stale; in general since we do change-by-amend, I also recommend that internal references within a PR should be done using "previous commit".

LGTM for the code otherwise.

---

Reviewed 1 of 2 files at r1, 1 of 3 files at r2.
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/vars.go, line 312 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16778#-KnkztuETIExpIDvV2pe:-Knnpw5fY2IzwCnTMAXq:b-5q2xkf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf8074a0ff54c5e84ca54f6159d5f7d44ff36a04/pkg/sql/vars.go#L312)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

at least when these functions are called directly by the statement interpreter, I think they'd be indeed be called on the "main goroutine". I don't know if there's more funky calls to these.
Anyway, since nobody cares about their performance, I'd lock the mutex and not worry about it.
</blockquote></details>

+1 on preventative lock and removing the TODO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16778#-:-Knnq18LFMUaB9ejwWAg:b-olgoo7)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
LGTM

When we do historical references, we should probably use a PR number, so we get a link and we can navigate to the discussion.

I talked to Andrew offline; the session is passed because some of those var functions need to change the session state.

---

Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16778#-:-KnoPbcQs6929B6LTzFG:bsndpi8)*
<!-- Sent from Reviewable.io -->

couchand committed:
sql: acquire read lock for session transaction

This change defensively acquires the read lock for accesses to the
session's transaction.  It's expected that references to session
variables will be made from the session's main goroutine, but since
there isn't a performance concern here, it seems like best practice.

couchand commented:
Both references have been changed to the relevant PRs to add additional context.  Thanks for your thoughts!

---

Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/vars.go, line 312 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16778#-KnkztuETIExpIDvV2pe:-KnofoycU4QzdgQJ_Gdl:b-tjaup) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf8074a0ff54c5e84ca54f6159d5f7d44ff36a04/pkg/sql/vars.go#L312)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

+1 on preventative lock and removing the TODO.
</blockquote></details>

Added the read locks here as well as to the one other reference that was previously assumed to be okay, following the same principle.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16778#-:-Knog90jSxNC8WVMk95K:b401jrk)*
<!-- Sent from Reviewable.io -->
